# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: 4053cb4ba9ad
# Full Hash: 4053cb4ba9ad573ee2e6dabdee56bed6e75eca52
# Author: Jeff Muizelaar <jmuizelaar@mozilla.com>
# Date: 2021-11-05 09:34:21
# Regressor Bug: 1540737
# File Overlap Count: 1
# Description:
#   Bug 1737263. Switch back to PushLayerWithBlend. r=mstange
#   
#   Bug 1540737 originally used PushLayerWithBlend here but PushLayerWithBlend is
#   not implemented for D2D. Since then, I believe we've stopped using D2D for
#   content drawing so it should be possible to switch back to PushLayerWithBlend.
# ==============================================================================

diff -r 38d0f40923e8 -r 4053cb4ba9ad layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Thu Nov 04 21:41:33 2021 +0000
+++ b/layout/painting/nsDisplayList.cpp	Thu Nov 04 21:42:14 2021 +0000
@@ -5027,37 +5027,12 @@
 
 void nsDisplayBlendMode::Paint(nsDisplayListBuilder* aBuilder,
                                gfxContext* aCtx) {
-  // This should be switched to use PushLayerWithBlend, once it's
-  // been implemented for all DrawTarget backends.
-  DrawTarget* dt = aCtx->GetDrawTarget();
-  int32_t appUnitsPerDevPixel = mFrame->PresContext()->AppUnitsPerDevPixel();
-  Rect rect = NSRectToRect(GetPaintRect(aBuilder, aCtx), appUnitsPerDevPixel);
-  rect.RoundOut();
-
-  // Create a temporary DrawTarget that is clipped to the area that
-  // we're going to draw to. This will include the same transform as
-  // is currently on |dt|.
-  RefPtr<DrawTarget> temp =
-      dt->CreateClippedDrawTarget(rect, SurfaceFormat::B8G8R8A8);
-  if (!temp) {
-    return;
-  }
-
-  RefPtr<gfxContext> ctx = gfxContext::CreatePreservingTransformOrNull(temp);
-
-  GetChildren()->Paint(aBuilder, ctx,
+  aCtx->GetDrawTarget()->PushLayerWithBlend(false, 1.0, nullptr,
+                                            mozilla::gfx::Matrix(), IntRect(),
+                                            false, BlendMode());
+  GetChildren()->Paint(aBuilder, aCtx,
                        mFrame->PresContext()->AppUnitsPerDevPixel());
-
-  // Draw the temporary DT to the real destination, applying the blend mode, but
-  // no transform.
-  temp->Flush();
-  RefPtr<SourceSurface> surface = temp->Snapshot();
-  gfxContextMatrixAutoSaveRestore saveMatrix(aCtx);
-  dt->SetTransform(Matrix());
-  dt->DrawSurface(
-      surface, Rect(surface->GetRect()), Rect(surface->GetRect()),
-      DrawSurfaceOptions(),
-      DrawOptions(1.0f, nsCSSRendering::GetGFXBlendMode(mBlendMode)));
+  aCtx->GetDrawTarget()->PopLayer();
 }
 
 gfx::CompositionOp nsDisplayBlendMode::BlendMode() {
